### PR TITLE
fix(resolvers): Fix custom resolvers on Windows

### PR DESCRIPTION
**Pull request by jakubadamw**
# Description

Fix npm/pnpm/yarn search on Windows.
Translate the main resolver module path in to use POSIX-style slashes.

# Type of change

- [ ] 💔 Breaking
- [ ] 🚀 Feature
- [X] 🐛 Fix
- [ ] 🛠️ Tooling
- [ ] 🔨 Refactoring
- [ ] 🧪 Test
- [ ] 📦 Dependency
- [ ] 📖 Requires documentation update

### CODE REVIEW

File: cli/crates/cli/tests/compilation_error.rs [hunk 1]
```diff
@@ -5,23 +5,25 @@ use std::collections::HashMap;
...
 
 #[test]
-#[cfg_attr(target_os = "windows", ignore)]
-fn compilation_error() {
+fn compilation_error_schema() {
     let mut env = Environment::init();
     env.grafbase_init(ConfigType::GraphQL);
     env.write_schema(COMPILATION_ERROR_SCHEMA);
+
+    std::thread::sleep(std::time::Duration::from_secs_f64(0.5));
```
Comment (yoav-lavi) [inline cli/crates/cli/tests/compilation_error.rs:19]: Why is this needed? Would be best to use something like our schema polling to be able to continue as fast as possible rather than waiting a set amount of time

File: cli/crates/cli/tests/compilation_error.rs [hunk 2]
```diff
@@ -37,12 +39,39 @@ fn compilation_error() {
 
     let response = client.gql::<Value>(DEFAULT_QUERY).send();
 
-    let errors: Option<Vec<String>> = dot_get_opt!(response, "errors");
+    let errors: Option<Vec<Value>> = dot_get_opt!(response, "errors");
```
Comment (yoav-lavi) [inline cli/crates/cli/tests/compilation_error.rs:42]: Are these not strings?

**Review by fbjork — APPROVED** [review]

File: cli/crates/cli/tests/compilation_error.rs [hunk 3]
```diff
@@ -52,20 +81,31 @@ fn compilation_error() {
...
     );
 
-    env.write_schema(COMPILATION_ERROR_RESOLVER_SCHEMA);
+    std::thread::sleep(std::time::Duration::from_secs_f64(0.5));
 
-    client.poll_endpoint_for_changes(30, 300);
+    // For now without watching before we investigate the issue.
+    env.grafbase_dev();
+
+    let client = env.create_client_with_longer_timeout().with_api_key();
```
Comment (yoav-lavi) [inline cli/crates/cli/tests/compilation_error.rs:89]: Maybe `client_with_options` and an options struct that lets you specify the timeout?

File: cli/crates/cli/tests/compilation_error.rs [hunk 4]
```diff
@@ -52,20 +81,31 @@ fn compilation_error() {
...
+    client.poll_endpoint(30, 300);
 
-    let response = client.gql::<Value>(COMPILATION_ERROR_RESOLVER_MUTATION).send();
+    let response = client.gql::<Value>(COMPILATION_ERROR_RESOLVER_QUERY).send();
 
-    let errors: Option<Vec<String>> = dot_get_opt!(response, "errors");
+    let errors: Option<Vec<Value>> = dot_get_opt!(response, "errors");
 
     assert_eq!(errors.map(|errors| !errors.is_empty()), Some(true));
```
Comment (yoav-lavi) [inline cli/crates/cli/tests/compilation_error.rs:97]: See above, sleeping for 2 secs in a test is a rather large increase to the overall runtime

File: cli/crates/server/src/bridge/resolvers.rs [hunk 5]
```diff
@@ -31,16 +32,109 @@ struct ResolverResponse {
...
+) -> Result<(tokio::task::JoinHandle<()>, u16), ApiError> {
+    let environment = Environment::get();
+
+    let resolver_worker_port = crate::servers::find_available_port_for_internal_use(main_worker_port)
+        .map_err(|_| ApiError::CouldNotFindPortForResolverWorker)?;
+    let resolver_worker_port_string = resolver_worker_port.to_string();
+
+    let miniflare_path = environment
+        .user_dot_grafbase_path
+        .join("node_modules/miniflare/dist/src/cli.js")
```
Comment (yoav-lavi) [inline cli/crates/server/src/bridge/resolvers.rs:76]: Could we use a const here?

File: cli/crates/server/src/custom_resolvers.rs [hunk 6]
```diff
@@ -99,39 +104,52 @@ async fn guess_package_manager_from_package_root(path: impl AsRef<Path>) -> Opti
     .next()
 }
 
+async fn extract_resolver_wrapper_worker_contents() -> Result<String, ResolverBuildError> {
+    trace!("extracting resolver wrapper worker contents");
+    let environment = Environment::get();
+    tokio::fs::read_to_string(
+        environment
+            .user_dot_grafbase_path
+            .join("custom-resolvers/wrapper-worker.js"),
```
Comment (yoav-lavi) [inline cli/crates/server/src/custom_resolvers.rs:113]: Could this be a const?

File: cli/crates/cli/tests/compilation_error.rs (hunk 2 above)
Comment (jakubadamw) [reply cli/crates/cli/tests/compilation_error.rs:42]: No. https://spec.graphql.org/October2021/#example-072c4

File: cli/crates/server/src/custom_resolvers.rs (hunk 6 above)
Comment (jakubadamw) [reply cli/crates/server/src/custom_resolvers.rs:113]: Yes, done!

File: cli/crates/server/src/bridge/server.rs [hunk 7]
```diff
@@ -121,14 +139,112 @@ async fn search_endpoint(
...
+                resolver_builds.insert(payload.resolver_name.clone(), ResolverBuild::Failed);
+                notify.notify_waiters();
+
+                let sender = handler_state.resolver_build_failure_sender.clone();
+                tokio::spawn(async move {
+                    // FIXME: We wait a second to avoid a race condition.
+                    // We want to allow the server to return a GraphQL reponse (with `errors`) before
+                    // we terminate the server to spawn the error server.
+                    // This will be fixed by removing this weird swap with a downtime window and introducing a single server.
+                    tokio::time::sleep(tokio::time::Duration::from_secs_f64(0.5)).await;
```
[the remark this answers is not in the document]
Comment (jakubadamw) [reply cli/crates/server/src/bridge/server.rs:236]: I deleted it in favour of not serving an error page for resolver compilation failures. If we want the latter, we'll need to do the single server refactoring first.

File: cli/crates/server/src/bridge/resolvers.rs (hunk 5 above)
Comment (jakubadamw) [reply cli/crates/server/src/bridge/resolvers.rs:76]: Done.

File: cli/crates/cli/tests/compilation_error.rs (hunk 3 above)
Comment (jakubadamw) [reply cli/crates/cli/tests/compilation_error.rs]: Done
